File: apps/web/src/lib/logging.ts
/**
 * Structured logger for the web service (ticket 0011).
 *
 * JSON to stdout via pino, level from `LOG_LEVEL` (default "info"). Every
 * context object is run through `redactPhi` (ADR 0008) so PHI can never land
 * in a log line by construction — pass records freely, the logger masks
 * member_id / name / dob / free-text before they leave the process.
 *
 * Always include correlation fields where available: request_id, clinic_id,
 * run_id. Use `log.child({ request_id, clinic_id })` per request.
 *
 * Server-only.
 */
import pino from "pino";
import { redactPhi } from "./phi";

const base = pino({
  level: process.env.LOG_LEVEL ?? "info",
  base: { service: "authmatic-web" },
  // Redact known PHI key paths defensively even if a caller forgets redactPhi.
  redact: {
    paths: ["member_id", "patient_name", "dob", "patient_ssn", "ssn", "justification", "rationale", "*.member_id", "*.patient_name", "*.dob"],
    censor: "[redacted]",
  },
  formatters: {
    level: (label) => ({ level: label }),
  },
});

type Context = Record<string, unknown>;

function emit(level: "info" | "warn" | "error", event: string, context?: Context): void {
  // Redact PHI keys in the supplied context before logging.
  base[level](context ? redactPhi(context) : {}, event);
}

export const log = {
  info: (event: string, context?: Context) => emit("info", event, context),
  warn: (event: string, context?: Context) => emit("warn", event, context),
  error: (event: string, context?: Context) => emit("error", event, context),
  /** Per-request child logger carrying correlation ids. */
  child: (bindings: Context) => base.child(redactPhi(bindings)),
};

File: apps/agent/src/logging.py
"""Structured JSON logging for the agent service (ticket 0011).

Dependency-free: stdlib `logging` + a JSON formatter to stdout, level from
`LOG_LEVEL`. Every line carries timestamp, level, service, logger, message,
and any correlation fields passed via `extra=` (request_id, clinic_id,
run_id) — e.g. loop.py's `planner.decision` events.

PHI must never be passed as a field value; use `src.phi.redact_phi` first
(ADR 0008). This formatter does not itself redact — it logs what it's given —
so callers redact at the call site, exactly as loop.py does.

`structlog` is a drop-in upgrade if richer processors are wanted later; the
call sites (`logging.getLogger(...).info(event, extra=...)`) don't change.
"""

from __future__ import annotations

import json
import logging
import os
from contextvars import ContextVar
from datetime import UTC, datetime

# Per-request correlation id (ticket 0021), set by the FastAPI middleware from
# the inbound X-Request-ID and surfaced on every log line within the request.
request_id_var: ContextVar[str | None] = ContextVar("request_id", default=None)

# Fields the stdlib LogRecord always has — everything else in __dict__ is a
# caller-supplied `extra` and gets surfaced as a structured field.
_STD = set(
    logging.makeLogRecord({}).__dict__.keys()
) | {"message", "asctime", "taskName"}


class JsonFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        payload = {
            "ts": datetime.fromtimestamp(record.created, tz=UTC).isoformat(),
            "level": record.levelname.lower(),
            "service": "authmatic-agent",
            "logger": record.name,
            "event": record.getMessage(),
        }
        rid = request_id_var.get()
        if rid and "request_id" not in record.__dict__:
            payload["request_id"] = rid
        for k, v in record.__dict__.items():
            if k not in _STD and not k.startswith("_"):
                payload[k] = v
        if record.exc_info:
            payload["exc"] = self.formatException(record.exc_info)
        return json.dumps(payload, default=str)


def setup_logging() -> None:
    """Install the JSON handler on the root logger. Idempotent."""
    level = os.environ.get("LOG_LEVEL", "INFO").upper()
    root = logging.getLogger()
    root.setLevel(level)
    # Replace any existing handlers so we don't double-log.
    for h in list(root.handlers):
        root.removeHandler(h)
    handler = logging.StreamHandler()
    handler.setFormatter(JsonFormatter())
    root.addHandler(handler)
